packages/imt.sol/contracts/internal/InternalLeanIMT.sol

```diff
@@ -103,15 +112,15 @@ library InternalLeanIMT {
         currentLevel = leaves;
 
         // Calculate the depth of the tree after adding the new values.
-        while (2 ** self.depth < self.size + leaves.length) {
+        while (2 ** self.depth < treeSize + leaves.length) {
             self.depth += 1;
         }
 
         // First index to change in every level.
-        uint256 currentLevelStartIndex = self.size;
+        uint256 currentLevelStartIndex = treeSize;
 
         // Size of the level used to create the next level.
-        uint256 currentLevelSize = self.size + leaves.length;
+        uint256 currentLevelSize = treeSize + leaves.length;
```

Comment:
Can we do the same for the `self.depth` in the `for()` below?

Reply:
I tried adding it when I was solving the issue, but it returns an error while compiling the contract:

```bash
CompilerError: Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.
```

Since it will avoid just one storage call, I decided to avoid adding that specific cache variable there.

Reply:
Damn, so unlucky with the stack limit. Makes sense, thank you!

Reply:
Hey @0xbok! Thanks.

The suggestion does not work for me, I get the same error. Is it working for you?

Reply:
@vplasencia this should fix the stack too deep error. 

```diff
-        while (2 ** self.depth < treeSize + leaves.length) {
-            self.depth += 1;
+        uint treeDepth = self.depth;
+        while (2 ** treeDepth < treeSize + leaves.length) {
+            ++treeDepth;
         }
+        self.depth = treeDepth;

         // First index to change in every level.
         uint256 currentLevelStartIndex = treeSize;
@@ -128,26 +130,26 @@ library InternalLeanIMT {
         // The size of the next level.
         uint256 nextLevelSize = ((currentLevelSize - 1) >> 1) + 1;

-        for (uint256 level = 0; level < self.depth; ) {
+        for (uint256 level = 0; level < treeDepth; ) {
             for (uint256 i = 0; i < numberOfNodes; ) {
-                uint256 rightNode;
                 uint256 leftNode;
-
-                // Assign the right node if the value exists.
-                if ((i + nextLevelStartIndex) * 2 + 1 < currentLevelSize) {
-                    rightNode = currentLevel[(i + nextLevelStartIndex) * 2 + 1 - currentLevelStartIndex];
-                }
-
                 // Assign the left node using the saved path or the position in the array.
                 if ((i + nextLevelStartIndex) * 2 < currentLevelStartIndex) {
                     leftNode = self.sideNodes[level];
                 } else {
                     leftNode = currentLevel[(i + nextLevelStartIndex) * 2 - currentLevelStartIndex];
                 }
+                uint256 rightNode;
+
+                // Assign the right node if the value exists.
+                if ((i + nextLevelStartIndex) * 2 + 1 < currentLevelSize) {
+                    rightNode = currentLevel[(i + nextLevelStartIndex) * 2 + 1 - currentLevelStartIndex];
+                }
```

Reply:
@0xbok 

Thanks, it works. Interesting, changing the order of the code.

Reply:
yeah, just had to remove the number of variables before the erroring line to reduce stack depth.